pydantic/schema.py

```diff
@@ -0,0 +1,392 @@
+    )
+    s.update(m_schema)
+    return s, m_definitions
+
+
+def model_type_schema(
+    class_: 'main.BaseModel',
+    *,
+    by_alias: bool,
+    model_name_map: Dict[Type['main.BaseModel'], str],
+    ref_prefix='#/definitions/',
+):
+    properties = {}
+    required = []
+    definitions = {}
+    for k, f in class_.__fields__.items():
+        f_schema, f_definitions = field_schema(
+            f, by_alias=by_alias, model_name_map=model_name_map, ref_prefix=ref_prefix
+        )
+        definitions.update(f_definitions)
+        if by_alias:
+            properties[f.alias] = f_schema
+            if f.required:
+                required.append(f.alias)
+        else:
+            properties[k] = f_schema
+            if f.required:
+                required.append(k)
+    out_schema = {'type': 'object', 'properties': properties}
+    if required:
+        out_schema['required'] = required
+    return out_schema, definitions
+
+
+def field_singleton_sub_fields_schema(
+    sub_fields: Sequence[Field],
+    *,
+    by_alias: bool,
+    model_name_map: Dict[Type['main.BaseModel'], str],
+    schema_overrides=False,
+    ref_prefix='#/definitions/',
+) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+    definitions = {}
+    if len(sub_fields) == 1:
+        return field_type_schema(
+            sub_fields[0],
+            by_alias=by_alias,
+            model_name_map=model_name_map,
+            schema_overrides=schema_overrides,
+            ref_prefix=ref_prefix,
+        )
+    else:
+        sub_field_schemas = []
+        for sf in sub_fields:
+            sub_schema, sub_definitions = field_type_schema(
+                sf,
+                by_alias=by_alias,
+                model_name_map=model_name_map,
+                schema_overrides=schema_overrides,
+                ref_prefix=ref_prefix,
+            )
+            definitions.update(sub_definitions)
+            sub_field_schemas.append(sub_schema)
+        return {'anyOf': sub_field_schemas}, definitions
+
+
+def field_singleton_schema(  # noqa: C901 (ignore complexity)
+    field: Field,
+    *,
+    by_alias: bool,
+    model_name_map: Dict[Type['main.BaseModel'], str],
+    schema_overrides=False,
+    ref_prefix='#/definitions/',
+) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+    definitions = {}
+    if field.sub_fields:
+        return field_singleton_sub_fields_schema(
+            field.sub_fields,
+            by_alias=by_alias,
+            model_name_map=model_name_map,
+            schema_overrides=schema_overrides,
+            ref_prefix=ref_prefix,
+        )
+    if field.type_ is Any:
+        return {}, definitions  # no restrictions
+    f_schema = {}
+    if issubclass(field.type_, Enum):
+        f_schema.update({'enum': [item.value for item in field.type_]})
+        # Don't return immediately, to allow adding specific types
+    # For constrained strings
+    if hasattr(field.type_, 'min_length'):
```

Comment:
a lot of this could be done with a dict rather than loads of ifs:

```py
attribute_lookup (better name?) = {'min_length': 'minLength', ...}

for field_name, schema_name in attribute_lookup.items():
    field_value = getattr(field.type_, 'field_name', None)
    if field_value is not None:
        f_schema[schema_name] =  field_value
```

Reply:
Done